osbs/build/user_params.py

```diff
@@ -0,0 +1,207 @@
+"""
+Copyright (c) 2018 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+from __future__ import print_function, absolute_import, unicode_literals
+
+import logging
+import json
+
+from osbs.build.spec import BuildParam, BuildIDParam, UserParam, BuildCommon
+from osbs.constants import DEFAULT_GIT_REF, REACTOR_CONFIG_ARRANGEMENT_VERSION
+from osbs.exceptions import OsbsValidationException
+from osbs.utils import get_imagestreamtag_from_image, make_name_from_git
+from osbs.repo_utils import AdditionalTagsConfig
+
+
+logger = logging.getLogger(__name__)
+
+
+class BuildUserParams(BuildCommon):
+    def __init__(self, build_json_dir=None):
+        # defines image_tag, koji_target, filesystem_koji_task_id, platform, arrangement_version
+        super(BuildUserParams, self).__init__()
+        self.build_json_dir = BuildParam('build_json_dir', default=build_json_dir)
+        self.git_uri = BuildParam('git_uri')
+        self.git_ref = BuildParam('git_ref', default=DEFAULT_GIT_REF)
+        self.git_branch = BuildParam('git_branch')
+        self.user = UserParam()
+        self.component = BuildParam('component')
+        self.name = BuildIDParam()
+        self.build_image = BuildParam('build_image')
+        self.build_imagestream = BuildParam('build_imagestream')
+        self.build_from = BuildParam('build_from')
+        self.build_type = BuildParam('build_type')
+        self.trigger_imagestreamtag = BuildParam('trigger_imagestreamtag')
+        self.imagestream_name = BuildParam('imagestream_name')
+        self.platforms = BuildParam("platforms", allow_none=True)
+        self.release = BuildParam("release", allow_none=True)
+        self.isolated = BuildParam("release", allow_none=True)
+        self.scratch = BuildParam("release", allow_none=True)
+        self.koji_task_id = BuildParam("koji_task_id", allow_none=True)
+        self.koji_parent_build = BuildParam("koji_parent_build", allow_none=True)
+        self.flatpak = BuildParam("flatpak", default=False)
+        self.flatpak_base_image = BuildParam("flatpak_base_image", allow_none=True)
+        self.signing_intent = BuildParam("signing_intent", allow_none=True)
+        self.yum_repourls = BuildParam("yum_repourls")
+        self.compose_ids = BuildParam("compose_ids", allow_none=True)
+        self.reactor_config_map = BuildParam("reactor_config_map", allow_none=True)
+        self.additional_tags = None
+        self.arrangement_version.value = REACTOR_CONFIG_ARRANGEMENT_VERSION
+        self.required_params = [
+            self.build_json_dir,
+            self.build_type,
+            self.git_ref,
+            self.git_uri,
+            self.koji_target,
+            self.user,
+        ]
+        self.convert_dict = {
```

Comment:
Can this be a tuple of strings instead? Less room to make a typo.
You can use `setattr` and `getattr` to modify/access attributes dynamically.

`convert_dict`/`convert_keys` should probably be a class attribute instead.

Reply:
I'm not at all sure what you mean here.

Reply:
Since you're populating `convert_dict` dynamically, my suggestion is no longer relevant.

What I had meant is that you don't really need a dict that maps to each BuildParam, all you need a list/tuple of attribute names.

```python
keys = ('koji_target', 'user')

# For reading attributes
for key in keys:
  getattr(self, key)

# For setting attributes
for key in keys:
  setattr(self, key, value)
```